keras/src/callbacks/orbax_checkpoint.py

```diff
@@ -0,0 +1,579 @@
+            # Create should_save_fn that saves every N steps
+            should_save_fn = (
+                lambda step, prev_step=None: step % save_interval == 0
+            )
+
+        # --- Orbax CheckpointManager Setup ---
+        from orbax.checkpoint import AsyncOptions
+
+        async_options = AsyncOptions(
+            timeout_secs=self.async_timeout_secs,
+            post_finalization_callback=self.post_finalization_callback,
+        )
+
+        options = ocp.CheckpointManagerOptions(
+            max_to_keep=max_to_keep,
+            keep_period=keep_period,
+            enable_async_checkpointing=save_on_background,
+            enable_background_delete=self.enable_background_delete,
+            async_options=async_options,
+            should_save_fn=should_save_fn,
+            save_decision_policy=save_decision_policy,
+        )
+        # Ensure directory exists (only needed on one process in multi-host)
+        if backend.get_process_index() == 0:
+            os.makedirs(directory, exist_ok=True)
+
+        # Create the CheckpointManager
+        self.manager = ocp.CheckpointManager(
+            directory=directory,
+            options=options,
+        )
+
+    def set_model(self, model):
+        self._model = model
+
+    def _should_save_on_batch(self, batch):
+        """Check if we should save on this batch."""
+        if self.save_freq == "epoch":
+            return False
+
+        if batch <= self._last_batch_seen:  # New epoch.
+            add_batches = batch + 1
+        else:
+            add_batches = batch - self._last_batch_seen
+        self._batches_seen_since_last_saving += add_batches
+        self._last_batch_seen = batch
+        self._total_batches_seen += add_batches
+
+        if self._batches_seen_since_last_saving >= self.save_freq:
+            self._batches_seen_since_last_saving = 0
+            return True
+        return False
+
+    def _get_current_step(self):
+        # A reliable way to get a global step count
+        # Using optimizer iterations is common
+        if hasattr(self.model, "optimizer") and hasattr(
+            self.model.optimizer, "iterations"
+        ):
+            # Convert potential backend tensor to int
+            return int(
+                backend.convert_to_numpy(self.model.optimizer.iterations)
+            )
+        else:
+            # Fallback: use global batch count
+            return self._total_batches_seen
+
+    def _save_checkpoint(self, step, logs=None):
+        """Save a checkpoint at the given step."""
+        if self.model is None:
+            return
+
+        # --- Prepare Composite State (Backend-Agnostic) ---
+        model_weights_np, optimizer_vars_np = _get_state_as_numpy(self.model)
+
+        if model_weights_np is None:
+            if self.verbose > 0:
+                print("OrbaxCheckpoint: Skipping save due to conversion error")
+            return
+
+        composite_state = {"model_weights": model_weights_np}
```

Comment:
If you use `get_state_tree` you won't need to do this anymore.

Reply:
ok